File: src/backend/app/routes/whatsapp.py
"""
FastAPI Routes para WhatsApp

Endpoints para generar links de WhatsApp con pedidos.
"""

from fastapi import APIRouter, HTTPException
from urllib.parse import quote
from typing import List
from dotenv import load_dotenv
import os

# Cargar variables de entorno
load_dotenv()

# Importar schemas (ajustar ruta si es necesario)
try:
    from app.models.schemas import CheckoutRequest, WhatsAppLinkResponse, CartItem
except ImportError:
    # Fallback para imports relativos
    import sys
    from pathlib import Path
    sys.path.insert(0, str(Path(__file__).parent.parent.parent))
    from app.models.schemas import CheckoutRequest, WhatsAppLinkResponse, CartItem

router = APIRouter(prefix="/api/whatsapp", tags=["whatsapp"])

# Número de teléfono del negocio (sin caracteres especiales)
WHATSAPP_BUSINESS_PHONE = os.getenv("WHATSAPP_BUSINESS_PHONE", "5493426158358")
BUSINESS_NAME = "Chocolatizados"


def calculate_total(items: List[CartItem]) -> float:
    """Calcular total del pedido."""
    return sum(item.price * item.quantity for item in items)


def format_cart_message(
    customer_name: str,
    customer_phone: str,
    customer_email: str,
    items: List[CartItem],
    delivery_address: str,
    total_price: float,
    notes: str = None
) -> str:
    """
    Genera mensaje de pedido para WhatsApp.
    
    Template estándar de Chocolatizados.
    """
    
    message = f"""Hola! 🍫 Nuevo pedido en {BUSINESS_NAME}

👤 Cliente: {customer_name}
📞 Teléfono: {customer_phone}
📧 Email: {customer_email}

📦 Pedido:"""
    
    # Items del carrito
    for idx, item in enumerate(items, 1):
        subtotal = item.price * item.quantity
        message += f"\n  {idx}. {item.name}"
        message += f"\n     Cantidad: {item.quantity} x ${item.price:.2f} = ${subtotal:.2f}"
        
        # Personalizaciones
        if item.allows_customization and item.customization_text:
            message += f"\n     ✨ Personalización: {item.customization_text}"
    
    message += f"\n\n💰 Total: ${total_price:.2f}"
    
    # Dirección de entrega
    if delivery_address:
        message += f"\n📍 Dirección: {delivery_address}"
    
    # Notas adicionales
    if notes:
        message += f"\n📝 Notas: {notes}"
    
    message += "\n\n✅ Por favor confirma el pedido por este medio"
    message += "\n\nLo que quieras decir decilo con chocolates 🍫"
    
    return message


@router.post("/generate-link", response_model=WhatsAppLinkResponse)
async def generate_whatsapp_link(checkout: CheckoutRequest) -> WhatsAppLinkResponse:
    """
    Genera link de WhatsApp con pedido pre-rellenado.
    
    Flujo:
    1. Frontend envía datos del carrito + cliente
    2. Backend genera mensaje formateado
    3. Backend retorna URL de WhatsApp
    4. Frontend abre link en nueva pestaña
    5. Usuario confirma manualmente en WhatsApp
    
    Nota: Este endpoint NO persiste datos en BD.
    """
    
    try:
        # Validar que haya items
        if not checkout.items:
            raise HTTPException(status_code=400, detail="El carrito está vacío")
        
        # Calcular total
        total_price = calculate_total(checkout.items)
        
        # Generar mensaje
        message = format_cart_message(
            customer_name=checkout.customer_name,
            customer_phone=checkout.customer_phone,
            customer_email=checkout.customer_email,
            items=checkout.items,
            delivery_address=checkout.delivery_address or "No especificada",
            total_price=total_price,
            notes=checkout.notes
        )
        
        # URL encode del mensaje
        encoded_message = quote(message)
        
        # Link de WhatsApp
        whatsapp_link = f"https://wa.me/{WHATSAPP_BUSINESS_PHONE}?text={encoded_message}"
        
        return WhatsAppLinkResponse(
            whatsapp_link=whatsapp_link,
            message_preview=message,
            total_price=total_price
        )
    
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Error generando link WhatsApp: {str(e)}"
        )


@router.get("/test")
async def test_whatsapp():
    """
    Endpoint de prueba para verificar configuración.
    """
    return {
        "status": "ok",
        "business_phone": WHATSAPP_BUSINESS_PHONE,
        "business_name": BUSINESS_NAME,
        "message": "Chocolatizados WhatsApp Integration OK"
    }

File: src/frontend/utils/whatsapp.ts
/**
 * Utilidad: generateWhatsAppLink
 * 
 * Genera un link de WhatsApp con el mensaje del pedido pre-rellenado
 * Se puede usar en Frontend (sin backend) o llamando al backend
 */

import { CartItem } from "@/hooks/useCart";

export interface CheckoutData {
  customer_name: string;
  customer_phone: string;
  customer_email: string;
  delivery_address?: string;
  items: CartItem[];
  notes?: string;
}

const BUSINESS_PHONE = "5493426158358"; // +54 9 3426 15-8358
const BUSINESS_NAME = "Chocolatizados";

/**
 * Genera el mensaje de pedido para WhatsApp
 */
export const generateOrderMessage = (data: CheckoutData): string => {
  const { customer_name, customer_phone, customer_email, items, delivery_address, notes } = data;

  let message = `Hola! 🍫 Nuevo pedido en ${BUSINESS_NAME}\n\n`;
  message += `👤 Cliente: ${customer_name}\n`;
  message += `📞 Teléfono: ${customer_phone}\n`;
  message += `📧 Email: ${customer_email}\n\n`;

  message += `📦 Pedido:\n`;
  items.forEach((item, idx) => {
    const subtotal = item.price * item.quantity;
    message += `  ${idx + 1}. ${item.name}\n`;
    message += `     Cantidad: ${item.quantity} x $${item.price.toFixed(2)} = $${subtotal.toFixed(2)}\n`;

    if (item.allows_customization && item.customization_text) {
      message += `     ✨ Personalización: ${item.customization_text}\n`;
    }
  });

  const total = items.reduce((sum, item) => sum + item.price * item.quantity, 0);
  message += `\n💰 Total: $${total.toFixed(2)}\n`;

  if (delivery_address) {
    message += `📍 Dirección: ${delivery_address}\n`;
  }

  if (notes) {
    message += `📝 Notas: ${notes}\n`;
  }

  message += `\n✅ Por favor confirma el pedido por este medio`;
  message += `\n\nLo que quieras decir decilo con chocolates 🍫`;

  return message;
};

/**
 * Genera el link de WhatsApp con el mensaje
 */
export const generateWhatsAppLink = (data: CheckoutData): string => {
  const message = generateOrderMessage(data);
  const encodedMessage = encodeURIComponent(message);
  return `https://wa.me/${BUSINESS_PHONE}?text=${encodedMessage}`;
};

/**
 * Abre el link de WhatsApp en una nueva ventana
 */
export const openWhatsAppLink = (data: CheckoutData): void => {
  const link = generateWhatsAppLink(data);
  window.open(link, "_blank", "width=600,height=700");
};

/**
 * (Opcional) Llama al backend para generar el link
 * Solo si quieres logging/analytics del servidor
 */
export const generateWhatsAppLinkFromBackend = async (
  data: CheckoutData,
  baseUrl: string = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000"
): Promise<string> => {
  const response = await fetch(`${baseUrl}/api/whatsapp/generate-link`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  });

  if (!response.ok) {
    throw new Error("Error generando link de WhatsApp");
  }

  const result = await response.json();
  return result.whatsapp_link;
};

/**
 * Genera mensaje SIMPLE de pedido para WhatsApp (Formato usuario)
 */
export const generateSimpleOrderMessage = (items: CartItem[], total: number): string => {
  // Emojis en Unicode para asegurar compatibilidad y evitar signos de pregunta
  const EMOJI_WAVE = "\uD83D\uDC4B";      // 👋
  const EMOJI_CLIPBOARD = "\uD83D\uDCCB"; // 📋
  const EMOJI_MONEY = "\uD83D\uDCB0";     // 💰
  const EMOJI_PERSON = "\uD83D\uDC64";    // 👤
  const EMOJI_CALENDAR = "\uD83D\uDCC5";  // 📅
  const EMOJI_CARD = "\uD83D\uDCB3";      // 💳

  let message = `Hola! ${EMOJI_WAVE} Quiero realizar el siguiente pedido:\n\n`;
  message += `${EMOJI_CLIPBOARD} Detalle:\n\n`;

  items.forEach((item) => {
    let desc = "";
    if (item.customization_text) {
      desc = ` (${item.customization_text})`;
    }

    message += `${item.quantity}x ${item.name}${desc}\n`;
  });

  message += `\n${EMOJI_MONEY} Total: $${total.toLocaleString('es-AR', { minimumFractionDigits: 2 })}\n\n`;

  message += `${EMOJI_PERSON} Nombre: \n`;
  message += `${EMOJI_CALENDAR} Fecha de retiro: \n`;
  message += `${EMOJI_CARD} Método de pago: \n`;

  message += `\nQuedo a la espera de confirmación!`;

  return message;
};

export const openSimpleWhatsAppLink = (items: CartItem[], total: number): void => {
  const message = generateSimpleOrderMessage(items, total);

  // Aseguramos que el mensaje esté codificado correctamente para URL (UTF-8)
  // Usamos api.whatsapp.com que suele manejar mejor los caracteres especiales que wa.me
  const encodedMessage = encodeURIComponent(message);
  const link = `https://api.whatsapp.com/send?phone=${BUSINESS_PHONE}&text=${encodedMessage}`;

  window.open(link, "_blank");
};
